pallets/collator-selection/src/migrations.rs

```diff
@@ -0,0 +1,96 @@
+// This file is part of Astar.
+
+// Copyright (C) Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use frame_support::{pallet_prelude::*, traits::OnRuntimeUpgrade};
+use frame_system::pallet_prelude::BlockNumberFor;
+use sp_std::marker::PhantomData;
+
+#[cfg(feature = "try-runtime")]
+use sp_std::vec::Vec;
+
+/// One-time migration that removes outdated LastAuthoredBlock entries.
+/// It keeps entries only for accounts that are currently:
+///   - active candidates
+///   - invulnerables
+///
+/// All other accounts are removed.
+pub struct LastAuthoredBlockCleanup<T: Config>(PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for LastAuthoredBlockCleanup<T> {
+    fn on_runtime_upgrade() -> Weight {
+        log::info!("Running CollatorSelection::LastAuthoredBlockCleanup...");
+
+        // Snapshot active identifiers for faster membership checks
+        let invulnerables = Invulnerables::<T>::get();
+        let candidate_accounts: sp_std::collections::btree_set::BTreeSet<T::AccountId> =
+            Candidates::<T>::get().into_iter().map(|c| c.who).collect();
+
+        let mut removed = 0u64;
+        let mut read = 0u64;
+        let mut write = 0u64;
+
+        // Use translate to selectively keep or drop keys
+        LastAuthoredBlock::<T>::translate::<BlockNumberFor<T>, _>(|account, old_value| {
```

Comment:
One suggestion I didn't submit - if you want to avoid `translate`, you can store keys in some vector, and then clean them up in another loop.

Reply:
I don't doubt it works, but it's purpose is to migrate from one type to another.